bouter/free/__init__.py

```diff
@@ -1,6 +1,250 @@
+from bouter import utilities, decorators
+import numpy as np
+import pandas as pd
 
 
 class FreelySwimmingExperiment(Experiment):
-    def get_bouts(self):
-        pass
+
+    @property
+    def n_tail_segments(self):
+        return self["tracking+fish_tracking"]["n_segments"] - 1
+
+    @property
+    def n_fish(self):
+        return self["tracking+fish_tracking"]["n_fish_max"]
+
+    @property
+    def camera_px_in_mm(self):
+        """ Return camera pixel size in millimeters
+
+        :param exp:
+        :return:
+        """
+        cal_params = self["stimulus"]["calibration_params"]
+        proj_mat = np.array(cal_params["cam_to_proj"])
+        return (
+            np.linalg.norm(np.array([1.0, 0.0]) @ proj_mat[:, :2]) * cal_params["mm_px"]
+        )
+
+    @property
+    def tail_columns(self):
+        """Return a nested list of names of columns with tracking data from all tracked segments.
+        One list for each fish tracked during the experiment.
+        """
+        return [[f"f{i}_theta_{j:02}" for j in range(self.n_tail_segments)] for i in range(self.n_fish)]
+
+
+    def _extract_bout(self, s, e, n_segments, i_fish=0, scale=1.0, dt=None):
+        bout = self._rename_fish(self.behavior_log.iloc[s:e], i_fish, n_segments)
+        # scale to physical coordinates
+        if dt is None:
+            dt = (bout.t.values[-1] - bout.t.values[0]) / bout.shape[0]
+
+        # pixels are scaled to millimeters (columns x, vx, y and vy)
+        bout.iloc[:, 1:5] *= scale
+        # velocities are additionally divided by the time difference to get mm/s
+        bout.iloc[:, 2:7:2] /= dt
+        return bout
+
+
+    def _fish_column_names(self, i_fish, n_segments):
+        return [
+                   "f{:d}_x".format(i_fish),
+                   "f{:d}_vx".format(i_fish),
+                   "f{:d}_y".format(i_fish),
+                   "f{:d}_vy".format(i_fish),
+                   "f{:d}_theta".format(i_fish),
+                   "f{:d}_vtheta".format(i_fish),
+               ] + ["f{:d}_theta_{:02d}".format(i_fish, i) for i in range(n_segments)]
+
+
+    def _fish_renames(self, i_fish, n_segments):
+        return dict(
+            {
+                "f{:d}_x".format(i_fish): "x",
+                "f{:d}_vx".format(i_fish): "vx",
+                "f{:d}_y".format(i_fish): "y",
+                "f{:d}_vy".format(i_fish): "vy",
+                "f{:d}_theta".format(i_fish): "theta",
+                "f{:d}_vtheta".format(i_fish): "vtheta",
+            },
+            **{
+                "f{:d}_theta_{:02d}".format(i_fish, i): "theta_{:02d}".format(i)
+                for i in range(n_segments)
+            }
+        )
+
+
+    def _rename_fish(self, df, i_fish, n_segments):
+        return df.filter(["t"] + self._fish_column_names(i_fish, n_segments)).rename(
+            columns=self._fish_renames(i_fish, n_segments)
+        )
+
+
+    @decorators.cache_results(cache_filename="behavior_log")
+    def compute_velocity(
```

Comment:
As written below, I actually think this function does not need to be cached, and the median filtering can be a part of it